### PR TITLE
WIP: Remove absolute urls (step 1)

### PR DESCRIPTION
**Pull request by binary1230**
This stuff is ready to merge though it's not even close to all the places we need to change in ubersystem to get things working purely off relative URLs.  This makes the uber app not be concerned with things like is it running over HTTPS vs HTTP (the proxy should worry about that)

More specific goals:
1) Code that generates HTML, or templates (except for email sending, cause I don't think there's a way around that) should ever reference the following vars:
```
path = "/uber"
hostname = "localhost"
url_root = "https://localhost:4443"
```

2) Anything that needs to reference another URL resource should either:
a) use relative paths
b) or look at the query string to figure out how to build an absolute URL (cherrypy already does this internally for redirects, check out ```cherrypy.url()``` which we can use elsewhere).  

some other info: http://tools.cherrypy.org/wiki/RoutesUrlGeneration

### CODE REVIEW

**Comment by binary1230** [comment]
I'm also opening this PR so you guys can get eyes on it early as it'll touch a bunch of UI stuff

File: uber/utils.py [hunk 1]
```diff
@@ -14,8 +14,6 @@ def __init__(self, page, *args, **kwargs):
         args = [self.quote(s) for s in args]
         kwargs = {k: self.quote(v) for k, v in kwargs.items()}
         cherrypy.HTTPRedirect.__init__(self, page.format(*args, **kwargs))
```
Comment (EliAndrewC) [inline uber/utils.py]: Hmm, I don't remember offhand why this if statement was here.  Are we 100% sure that removing it isn't going to cause a problem?

My vague recollection is that since CherryPy itself is speaking HTTP, we had to replace that with HTTPS since we're behind Apache.  Or something like that, I don't really recall.  (This is a good reason to get better about writing docstrings, haha.)

File: uber/server.py [hunk 2]
```diff
@@ -52,7 +52,7 @@ def default(self, *path_args, **kwargs):
 @all_renderable()
 class Root:
     def index(self):
```
Comment (EliAndrewC) [inline uber/server.py]: +1 on moving this stuff into ``common`` to avoid this problem.

File: uber/site_sections/common.py [hunk 3]
```diff
@@ -0,0 +1,7 @@
+from uber.common import *
+
+
+@all_renderable()
+class Root:
+    def index(self):
+        return render('index.html')
```
Comment (EliAndrewC) [inline uber/site_sections/common.py]: It would be more idiomatic to just move the ``index.html`` template into the ``uber/templates/common`` directory and then we could just say ``return {}`` and not need to specify the filename.

File: uber/utils.py (hunk 1 above)
Comment (binary1230) [inline uber/utils.py]: I think removing this is going to be fine, but I'll find out in testing. I'd like to remove it if possible.

I *think* the only downside is that a request coming in via non-SSL will get one additional 302 redirect by the external proxy.  I think that's a good tradeoff here in order to make our code a little more robust in not having to know it's own URL, and that there's only two scenarios where we need this:
1) User managed to reach ubersystem port by not using HTTP (which shouldn't be possible because that should get redirected first by nginx to HTTPS)
2) Internally, we're somehow building an absolute redirect URL to ourselves that includes HTTP (which, I don't think we'd ever do), but even if we did, the proxy server should redirect that request as well.

File: uber/site_sections/common.py (hunk 3 above)
Comment (binary1230) [inline uber/site_sections/common.py]: good call. i'll change that.

File: uber/utils.py (hunk 1 above)
Comment (binary1230) [inline uber/utils.py]: So hey I think I figured out why it was there, it was part of the http_to_https.py file which relied on it to do the actual redirection of http->https.

i.e. stuff that should likely go away.
@EliAndrewC definitely would like your eyes on this bit.  so far, it didn't look like anything referenced it so looks OK to delete this and the other stuff in https://github.com/magfest/ubersystem/commit/94fe1053820e51e4647b2e6e340583bd321fb92d

File: uber/site_sections/common.py (hunk 3 above)
Comment (binary1230) [inline uber/site_sections/common.py]: fixed

File: uber/site_sections/preregistration.py [hunk 4]
```diff
@@ -67,7 +67,11 @@ def unpaid_preregs(self):
     def paid_preregs(self):
         return cherrypy.session.setdefault('paid_preregs', [])
```
Comment (EliAndrewC) [inline uber/site_sections/preregistration.py]: Why this change?  The main reason to pass ``None`` and then instantiate a thing is if the argument is mutable.  For example, it's generally bad form to say ``def foo(bar=[])`` because ``[]`` is mutable and so the default value could theoretically change between invocations.  That's not really the case here (though technically I suppose someone could mutate the object if they went out of their way to do so).  I don't have any particular objection to it (though it does make the code slightly more complicated), but what was the reason for this change?

File: uber/http_to_https.py [hunk 5]
```diff
@@ -1,13 +0,0 @@
-from uber.common import *
```
Comment (EliAndrewC) [inline uber/http_to_https.py]: +1 on removing this.

Although for future reference, we should put our comments on the Files Changed tab of the PR rather than in the commit, since the comments which @binary1230 made do not show up here due to being left on the commit itself.

File: uber/utils.py (hunk 1 above)
Comment (EliAndrewC) [inline uber/utils.py]: Gotcha, yeah that makes sense.

+1 on this change.

File: uber/site_sections/preregistration.py (hunk 4 above)
Comment (binary1230) [inline uber/site_sections/preregistration.py]: ok yea this is very non-obvious and I only caught it in the debugger.  the issue I'm fixing is real, though, there might be a more pythonic way to deal with this.

# how HTTPRedirect works:
deep down in cherrypy's HTTPRedirect's __init__, it attempts to take our relative URL and create an absolute URL from it by looking at the request data from the browser.

i.e. if you accessed cherrypy with a request string from your browser like 'http://uberserver.com:765/uber/preregistration/someplace' and you ask to redirect to 'index', cherrypy will mix those two things together and generate: 'http://uberserver.com:765/uber/preregistration/index'

likewise, if the way you accessed ubersystem was different (let's say you visited the page through another proxy server somehow) it still generates the correct result.  

If you later accessed this same ubersystem instance via 'http://localhost8765/uber/preregistration/something' and redirected to 'coolestpageever', cherrypy will look at your request string and build the following URL: 'http://localhost8765/uber/preregistration/coolestpageever'

TLDR: Cherrypy creates a base URL by looking at the query string your browser sent it.

So, HTTPRedirect() only does that generation on __init__() when it's created, and it needs to have the context i.e. access to THE CURRENT request string from the browser.

# Why the current implementation might be an issue

This was really non-obvious and I only caught it by accident in the debugger.

When we did had this function definition:

```python
def _get_unsaved(self, id, if_not_found=HTTPRedirect('index')):
```

The *definition* was creating a HTTPRedirect object ```at import time``` (i.e. at sideboard start). 

This HTTPRedirect instance would not have any request context to look at (since this code is not being executed as part of filling a browser request)

Cherrypy doesn't crash on this, but instead of building an absolute path using the querystring, it just kind of guesses based on config data.  The comments even say this might cause issues:
```python
# No request.app (we're being called outside a request).
# We'll have to guess the base from server.* attributes.
# This will produce very different results from the above
# if you're using vhosts or tools.proxy.
```

(that code is in __init__.py in cherrypy) in the function ```def url(path="", qs="", script_name=None, base=None, relative=None):```

# short version

Putting HTTPRedirect() in the function definition makes a global object that we can later 'raise' if needed.  The problem is it will not look at a querystring, and thus, might fail in various configurations of proxy stuff.

My goal was to make it so that no matter how your browser ended up getting to cherrypy's port, you get the right data back out of it with minimal tricks.

So my fix here is simply: We shouldn't create HTTPRedirect objects and save them in a global state (which is what the original code did), we should only create them right before they're needed.

Comment (binary1230) [inline uber/site_sections/preregistration.py]: and btw, I think that in all production versions, this just happened to work OK by happy coincidence.

in the future, we might have multiple ways to access ubersystem (i.e. like what we're almost moving towards with making uber accessible at-even both on-site and on the internet), we will need this to function correctly no matter how your request came in.

extra sidenote: I only discovered this when I had a breakpoint on HTTPRedirect::__init__() and it was going off at server start.  originally, I thought maybe the prereg form was doing some kind of javascript automated check, but it happened even if I closed all my browser windows.  was pretty surprised to find that it was being created as a global object (I swear man python every day I learn something totally crazy about that language, it makes sense though)

File: uber/http_to_https.py (hunk 5 above)
Comment (binary1230) [inline uber/http_to_https.py]: yea dag i was hoping they'd magic up over here.

File: uber/site_sections/preregistration.py (hunk 4 above)
Comment (EliAndrewC) [inline uber/site_sections/preregistration.py]: Gotcha, that all makes sense, thanks for the detailed explanation.

**Comment by binary1230** [comment]
hey, will be pushing to get this merged into master today as long as that doesn't affect Anthrocon.  i've been using this branch heavily with the docker stuff, but hadn't tested it as thoroughly with the puppet deploy.  it -should- all be fine, will test a bit more.  @EliAndrewC @kitsuta gimme a +1 if OK to merge, and I'd like to pull the trigger on this one

**Comment by kitsuta** [comment]
Haha whoops, sorry for not reviewing this yesterday, I couldn't find it because it still had "WIP" in the title.

Anyway looks good, ship it.

**Comment by EliAndrewC** [comment]
+1 on merging, I haven't tested it locally or anything, but the code looks good to me.  Since both Vicki and myself have signed off, I'll go ahead and hit merge now.

**Comment by binary1230** [comment]
Uhhh crap it wasn't ready to go woops, vagrant needs to do some port forwarding stuff

Cherrypy tools.proxy stuff

**Comment by binary1230** [comment]
Works fine with docker, doesn't work if port is weird like 4443

**Comment by EliAndrewC** [comment]
Alas.  So are we still using Vagrant for development for the forseeable future?  I've been meaning to ask about that, since I've still been using my existing Vagrant dev setup, but once I start jumping into MagClassic stuff this week, I should probably try using whatever everyone else is using.

**Comment by binary1230** [comment]
Yea, vagrant is still the only officially supported methods right now. The various docker implementations aren't ready for prime time just yet and not for classic

**Comment by binary1230** [comment]
so hey got the fix for this, it's actually not any of this code, it's a missing nginx proxy_redirect that handles when uber wants to do a 'Location' header (i.e. a redirect), it needs to be re-written.  Will be checking that in in a few.  but, this could is now tested and correct as far as I know.

if anyone hits anything weird with redirects  or URLs or something, let me know.

**Comment by binary1230** [comment]
Also you will need to recreate your vagrant setup from scratch as a lot has changed recently in there
